README.md
# vue-snotify

[![Build Status](https://travis-ci.org/artemsky/vue-snotify.svg?branch=master)](https://travis-ci.org/artemsky/vue-snotify)
[![NPM Version](https://img.shields.io/npm/v/vue-snotify.svg)](https://www.npmjs.com/package/vue-snotify)
[![NPM Downloads](https://img.shields.io/npm/dt/vue-snotify.svg)](https://www.npmjs.com/package/vue-snotify)
[![Dev dependencies status list](https://david-dm.org/artemsky/vue-snotify/dev-status.svg)](https://david-dm.org/artemsky/vue-snotify?type=dev)

## Example
https://artemsky.github.io/vue-snotify/

## Features

- 9 types of toast notifications (async, confirm, prompt, html and more...)
- Many config options (icons, backdrop, timeout, position and much more...)
- Repository includes 3 different styles. So you can use on of them, or create your own.
- Callbacks
- 5KB minified and gzipped
- Typescript definitions
- ESM, CommonJS, UMD versions

###### Looking for an Angular version? [Here](https://github.com/artemsky/ng-snotify/)

![Snotify Gif](https://artemsky.github.io/vue-snotify/static/vue-snotify-demo.gif)

## Installation

###### NPM 5
`npm install vue-snotify`

###### yarn
`yarn add vue-snotify`

[see detailed instruction](https://artemsky.github.io/vue-snotify/documentation/installation.html)

## Documentation and Examples

Documentation - [here](https://artemsky.github.io/vue-snotify/documentation/index.html)  
Example application source - [here](https://github.com/artemsky/vue-snotify/tree/master/example/src)   
Change Log - [here](https://github.com/artemsky/vue-snotify/blob/master/CHANGELOG.md)

#### CDN
  - JS
    - [unpkg.com](https://unpkg.com/vue-snotify@latest/vue-snotify.min.js) | [jsDelivr.com](https://cdn.jsdelivr.net/npm/vue-snotify@latest/vue-snotify.min.js)
  - CSS
    - material [unpkg.com](https://unpkg.com/vue-snotify@latest/styles/material.css) | [jsDelivr.com](https://cdn.jsdelivr.net/npm/vue-snotify@latest/styles/material.css)
    - simple [unpkg.com](https://unpkg.com/vue-snotify@latest/styles/simple.css) | [jsDelivr.com](https://cdn.jsdelivr.net/npm/vue-snotify@latest/styles/simple.css)
    - dark [unpkg.com](https://unpkg.com/vue-snotify@latest/styles/dark.css) | [jsDelivr.com](https://cdn.jsdelivr.net/npm/vue-snotify@latest/styles/dark.css)

## License

MIT © [artemsky](mailto:mr.artemsky@gmail.com)


CHANGELOG.md
## Change Log

### v3.2.2
  - fix build process #66
  - add fix backdrop #57

### v3.2.1
  - remove `console.log` [#48](https://github.com/artemsky/vue-snotify/issues/48)
  
### v3.2.0
  - stop buttons event propagation [#37](https://github.com/artemsky/vue-snotify/issues/37)
  - added ability to hide icon`icon: false` [#28](https://github.com/artemsky/vue-snotify/issues/28)
  - back to SFC. Now with runtime-only build [#14](https://github.com/artemsky/vue-snotify/issues/14)

### v3.1.0
  - add variables namespace [#30](https://github.com/artemsky/vue-snotify/issues/30)
  - oneAtTime | preventDuplicates - add global config options [#28](https://github.com/artemsky/vue-snotify/issues/28)
  - add button custom className [#24](https://github.com/artemsky/vue-snotify/issues/24)
  - fix toastChanged event initialization [#20](https://github.com/artemsky/vue-snotify/issues/20)
  - bump dependencies

### v3.0.4
  - fix icons [#17](https://github.com/artemsky/vue-snotify/issues/15)
### v3.0.3
  - fix icons [#15](https://github.com/artemsky/vue-snotify/issues/15)
### v3.0.2
  - fix backdrop [#16](https://github.com/artemsky/vue-snotify/issues/16)
### v3.0.1
  - improve documentation
  - escape SVG icons for Safari, Firefox
### v3.0.0
- **Breaking changes -** [Look migration guide](https://artemsky.github.io/vue-snotify/documentation/essentials/upgrade.html)
- **Features**
  - [New Documentation](https://artemsky.github.io/vue-snotify/documentation) based on gitbook
  - Prompt input [validation](https://artemsky.github.io/vue-snotify/documentation/essentials/examples.html#prompt--validation)
  - New [callbacks](https://artemsky.github.io/vue-snotify/documentation/api/callbacks.html) with jQuery-like syntax
  - Reworked toast arguments ordering based on - see [function signatures](https://artemsky.github.io/vue-snotify/documentation/api/snotify.html)
  - Replaced all animations and icons to styles (now it more customizable up to you)
  - Add toast [default configuration](https://artemsky.github.io/vue-snotify/documentation/api/options.html)
  - Reworked toast `timeout` interval - now based on `requestAnimationFrame`

### v2.0.3
  - bump dependencies
  - [Fix `action is not a function` error](https://github.com/artemsky/vue-snotify/pull/3) by @grantcarthew

### v2.0.0
  - optimize typings
  - **[BREAKING CHANGE]** As most of you wanted this. I moved `SnotifyService` onto `Vue.prototype`. Now accessible only from to `Vue.$snotify // this.$snotify`
  - create CDN versions
  - add VueJS like [documentation](https://artemsky.github.io/vue-snotify/documentation)
### v1.1.0
  - add Typescript typings
  
### v1.0.1
  - fix ES module name

### v1.0.0
  - Ported from [Angular version](https://github.com/artemsky/ng-snotify/)


gulpfile.js
/* eslint-disable */
const gulp = require('gulp'),
  path = require('path'),
  rollup = require('gulp-rollup'),
  rename = require('gulp-rename'),
  del = require('del'),
  sass = require('gulp-sass'),
  postcss = require('gulp-postcss'),
  autoprefixer = require('autoprefixer'),
  replace = require('rollup-plugin-replace'),
  cjs = require('rollup-plugin-commonjs'),
  node = require('rollup-plugin-node-resolve'),
  inject = require('gulp-inject-string'),
  uglify = require('rollup-plugin-uglify'),
  inlineResources = require('./tools/gulp/inline-resources'),
  ts = require("gulp-typescript"),
  merge = require('merge2'),
  greplace = require('gulp-replace'),
  dMerge = require('deepmerge'),
  typescript = require('rollup-plugin-typescript2'),
  VuePlugin = require('rollup-plugin-vue').default;

const {version, license, author, name} = require('./package.json');
const banner =
  '/**\n' +
  ' * ' + name + ' v' + version + '\n' +
  ' * (c) ' + new Date().getFullYear() + ' ' + author.name + ' <' + author.email + '>\n' +
  ' * @license ' + license + '\n' +
  ' */\n';

const rootFolder = path.join(__dirname);
const srcFolder = path.join(rootFolder, 'src');
const tmpFolder = path.join(rootFolder, '.tmp');
const distFolder = path.join(rootFolder, 'dist');
const buildFolder = path.join(rootFolder, 'build');
const tsProject = ts.createProject( {
      "declaration": true,
      "module": "es2015",
      "target": "es5",
      "baseUrl": ".",
      "stripInternal": true,
      "emitDecoratorMetadata": true,
      "experimentalDecorators": true,
      "moduleResolution": "node",
      "outDir": "./build",
      "rootDir": ".",
      "lib": [
        "es2015",
        "dom"
      ],
      "include": [
        "src/sfc.d.ts"
      ],
      "skipLibCheck": true
    }
);

/**
 * 1. Delete /dist folder
 */
gulp.task('clean:dist', function () {

  // Delete contents but not dist folder to avoid broken npm links
  // when dist directory is removed while npm link references it.
  return deleteFolders([distFolder + '/**', '!' + distFolder]);
});

/**
 * 2. Clone the /src folder into /.tmp. If an npm link inside /src has been made,
 *    then it's likely that a node_modules folder exists. Ignore this folder
 *    when copying to /.tmp.
 */
gulp.task('copy:source', function () {
  return gulp.src([`${srcFolder}/**/*`, `!${srcFolder}/node_modules`])
    .pipe(gulp.dest(tmpFolder));
});

/**
 * 3. Inline template (.html) and style (.css) files into the the component .ts files.
 *    We do this on the /.tmp folder to avoid editing the original /src files
 */
gulp.task('inline-resources', function () {
  return Promise.resolve()
    .then(() => inlineResources(tmpFolder));
});

/**
 * 4. Run the typescript compiler, on the /.tmp folder. This will output all
 *    compiled modules to the /build folder.
 */

gulp.task('typescript', function () {
  const tsResult = gulp.src(`${tmpFolder}/**/*.ts`)
  .pipe(tsProject());

  return tsResult.dts.pipe(gulp.dest(buildFolder));
});


/**
 * 5. Replace static keyword in /.build declarations files? just because Vue interface doesn't accept that
 *    compiled modules to the /build folder.
 */

gulp.task('replaceService', function(){
  return gulp.src(`${buildFolder}/SnotifyService.d.ts`)
    .pipe(greplace('static', ''))
    .pipe(gulp.dest(buildFolder));
});

/**
 * 5. Inject Vue interface override
 */

gulp.task('injectDefinitions', function(){
  const fs = require('fs');
  return gulp.src(`${buildFolder}/index.d.ts`)
    .pipe(inject.prepend(fs.readFileSync(`${srcFolder}/vue.d.ts`, "utf8")))
    .pipe(gulp.dest(buildFolder));
});


/**
 * 6. Run rollup inside the /build folder to generate our Flat ES module and place the
 *    generated file into the /dist folder
 */
gulp.task('rollup:fesm', function () {
  return gulp.src(`${buildFolder}/**/*.js`)
  // transform the files here.
    .pipe(rollup(getRollupOptions({
      output: {
        format: 'es',
      },
    })))
    .pipe(inject.prepend(banner))
    .pipe(rename('vue-snotify.esm.js'))
    .pipe(gulp.dest(distFolder));
});

/**
 * 7. Run rollup inside the /build folder to generate our UMD module and place the
 *    generated file into the /dist folder
 */
gulp.task('rollup:umd', function () {
  return gulp.src(`${srcFolder}/**/*.ts`)
  // transform the files here.
    .pipe(rollup(getRollupOptions({
        output: {
          format: 'umd',
          exports: 'named',
          name: 'vue-snotify',
        },
      })))
    .pipe(inject.prepend(banner))
    .pipe(rename('vue-snotify.js'))
    .pipe(gulp.dest(distFolder));
});

/**
 * 8. Run rollup inside the /build folder to generate our UMD module and place the
 *    generated file into the /dist folder
 */
gulp.task('rollup:umd:min', function () {
  return gulp.src(`${buildFolder}/**/*.js`)
  // transform the files here.
    .pipe(rollup(getRollupOptions({
      output: {
        format: 'umd',
        exports: 'named',
        name: 'vue-snotify',
      },
      plugins: [
        uglify()
      ]
    })))
    .pipe(inject.prepend(banner))
    .pipe(rename('vue-snotify.min.js'))
    .pipe(gulp.dest(distFolder));
});

/**
 * 9. Run rollup inside the /build folder to generate our CommonJS module and place the
 *    generated file into the /dist folder
 */
gulp.task('rollup:cjs', function () {
  return gulp.src(`${tmpFolder}/**/*.js`)
  // transform the files here.
    .pipe(rollup(getRollupOptions({
      output: {
        format: 'cjs',
        exports: 'named',
        name: 'vue-snotify',
      }
    })))
    .pipe(inject.prepend(banner))
    .pipe(rename('vue-snotify.common.js'))
    .pipe(gulp.dest(distFolder));
});


/**
 * 10. Copy all the files from /build to /dist, except .js files. We ignore all .js from /build
 *    because with don't need individual modules anymore, just the Flat ES module generated
 *    on steps 6,7,8,9.
 */
gulp.task('copy:build', function () {
  return gulp.src([`${buildFolder}/**/*`, `!${buildFolder}/**/*.js`])
    .pipe(gulp.dest(distFolder));
});

/**
 * 11. Copy package.json from /src to /dist
 */
gulp.task('copy:manifest', function () {
  return gulp.src([`${srcFolder}/package.json`])
    .pipe(gulp.dest(distFolder));
});

/**
 * 12. Copy README.md from / to /dist
 */
gulp.task('copy:readme', function () {
  return gulp.src([path.join(rootFolder, 'README.md')])
    .pipe(gulp.dest(distFolder));
});

/**
 * 13. Delete /.tmp folder
 */
gulp.task('clean:tmp', function () {
  return deleteFolders([tmpFolder]);
});

/**
 * 14. Delete /build folder
 */
gulp.task('clean:build', function () {
  return deleteFolders([buildFolder]);
});

/**
 * 15. Compile styles into separate bundle
 */
gulp.task('styles:compile', function () {
  return gulp.src([`${srcFolder}/styles/*.scss`])
    .pipe(sass().on('error', sass.logError))
    .pipe(postcss([
      autoprefixer({browsers: ["last 2 versions", "safari >= 7"], cascade: false})
    ]))
    .pipe(gulp.dest(`${distFolder}/styles`));
});

/**
 * 16. Copy styles sources to our dist folder
 */
gulp.task('styles:copy', function () {
  return gulp.src([`${srcFolder}/styles/**`])
    .pipe(gulp.dest(`${distFolder}/styles`));
});


/**
 * 17. Copy manifest, readme, to our dist folder
 */
gulp.task('copy', gulp.parallel('copy:manifest', 'copy:readme'));

gulp.task('compile', gulp.series(
  'copy:source',
  'inline-resources',
  'typescript',
  'copy:build',
  gulp.parallel('rollup:fesm', 'rollup:umd', 'rollup:cjs', 'rollup:umd:min'),
  (done) => {
    console.log('LIBRARY: compilation finished successfully');
    done();
  }
));

/**
 * Watch for any change in the /src folder and compile files
 */
gulp.task('watch', function () {
  gulp.watch(`${srcFolder}/**/*`, gulp.parallel('compile'));
  gulp.watch(`${srcFolder}/styles/**`, gulp.parallel('styles:build'));
});

gulp.task('clean', gulp.parallel('clean:dist', 'clean:tmp', 'clean:build'));

gulp.task('styles:build', gulp.parallel('styles:compile', 'styles:copy', (done) => {
  console.log('STYLES: compilation finished successfully');
  done();
}));

gulp.task('build', gulp.series(
  'clean',
  'compile',
  'copy',
  'clean:build',
  'clean:tmp',
  'styles:build'
));

gulp.task('default', gulp.series(
  'build',
  'watch'
));


/**
 * Deletes the specified folderva
 * @param folders
 */
function deleteFolders(folders) {
  return del(folders);
}

/**
 * Generate configuration object for rollup
 * @returns {*}
 */
function getRollupOptions(options = {}) {
  return dMerge({
    // Bundle's entry point
    // See https://github.com/rollup/rollup/wiki/JavaScript-API#entry
    input: `${srcFolder}/index.ts`,

    // Allow mixing of hypothetical and actual files. "Actual" files can be files
    // accessed by Rollup or produced by plugins further down the chain.
    // This prevents errors like: 'path/file' does not exist in the hypothetical file system
    // when subdirectories are used in the `src` directory.
    allowRealFiles: true,

    // A list of IDs of modules that should remain external to the bundle
    // See https://github.com/rollup/rollup/wiki/JavaScript-API#external
    external: [
      'vue',
      'Vue'
    ],

    output: {
      // strict: true,
      // See https://github.com/rollup/rollup/wiki/JavaScript-API#globals
      globals: {
        typescript: 'ts',
        vue: 'Vue'
      },
    },


    plugins: [
      typescript({
        typescript: require('typescript'),
        tsconfigOverride: {
          "declaration": true,
          "module": "es2015",
          "target": "es5",
          "baseUrl": ".",
          "stripInternal": true,
          "emitDecoratorMetadata": true,
          "experimentalDecorators": true,
          "noImplicitAny": true,
          "moduleResolution": "node",
          "outDir": "./build",
          "rootDir": ".",
          "lib": [
            "es2015",
            "dom"
          ],
          "includes": [
            "src/sfc.d.ts",
            "src/options.d.ts"
          ],
          "files": [
            "src/sfc.d.ts",
            "src/options.d.ts"
          ],
          "skipLibCheck": true
        },
        cacheRoot: '.tmp/' + (Math.floor(Math.random() * (Date.now() - 1)) + 1)
      }),
      VuePlugin({
        defaultLang: { script: 'ts' },
      }),
      node({
        jsnext: true,
        module: true,
        main: true,  // for commonjs modules that have an index.js
        browser: true
      }),
      cjs({
        namedExports: {
          // left-hand side can be an absolute path, a path
          // relative to the current directory, or the name
          // of a module in node_modules
          'vue-property-decorator': [ 'Component' ]
        }
      }),
      replace({
        'process.env.NODE_ENV': JSON.stringify('production')
      }),
    ]

  }, options);
}
